### PR TITLE
Fix issues with file reading and writing with Python code

### PR DESCRIPTION
**Pull request by erik-megarad**
### Background

Several times I've had Auto-GPT fail to recognize "does not exist" error for files, and it will attempt other fixes such as analyzing the code (for a file that does not exist). I've also had issues with python code not being executed in the working directory, which means that code written by Auto-GPT that reads and writes files will usually not work (since they use relative paths).

These issues are related:
https://github.com/Significant-Gravitas/Auto-GPT/issues/4439
https://github.com/Significant-Gravitas/Auto-GPT/issues/4413
https://github.com/Significant-Gravitas/Auto-GPT/issues/4366
https://github.com/Significant-Gravitas/Auto-GPT/issues/4360
https://github.com/Significant-Gravitas/Auto-GPT/issues/3583

### Changes

- Change the error messages for file not found errors so that they're easier to recognize as such
- Set `cwd` when executing python code so that relative paths work as expected

### Documentation

- In-code comments

### Test Plan

- Wrote a new test for reading a file that does not exist (was previously untested)
- Wrote a new test for executing a python file that does not exist (was previously untested)

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes. 
- [x] I have run the following commands against my code to ensure it passes our linters:
    ```shell
    black .
    isort .
    mypy
    autoflake --remove-all-unused-imports --recursive --ignore-init-module-imports autogpt tests --in-place
    ```

### CODE REVIEW

**Comment by erik-megarad** [comment]
Example of the new error being recognized:

```
WEATHERGPT THOUGHTS:  I have found two reliable sources of weather data for Paris, Texas on October 10th, 1991. I will cross-reference the data from both sources to ensure accuracy and reliability.
REASONING:  Cross-referencing multiple sources of weather data will help ensure the accuracy and reliability of the information.
PLAN:
-  Retrieve the weather data from both sources.
-  Cross-reference the data to ensure accuracy and reliability.
-  Present the weather information in a clear and concise manner.
CRITICISM:  I need to be more efficient in cross-referencing multiple sources of weather data.
SPEAK:  I have found two reliable sources of weather data for Lilburn, Georgia on October 12th, 1991. Please wait a moment while I cross-reference the data to ensure accuracy and reliability.
NEXT ACTION:  COMMAND = execute_python_file ARGUMENTS = {'filename': '/app/autogpt/auto_gpt_workspace/cross_reference_weather_data.py'}
  Enter 'y' to authorise command, 'y -N' to run N continuous commands, 's' to run self-feedback commands, 'n' to exit program, or enter feedback for WeatherGPT...
  Asking user via keyboard...
Input:y
-=-=-=-=-=-=-= COMMAND AUTHORISED BY USER -=-=-=-=-=-=-=
  Executing python file '/app/autogpt/auto_gpt_workspace/cross_reference_weather_data.py' in working directory '/app/autogpt/auto_gpt_workspace'
SYSTEM:  Command execute_python_file returned: python: can't open file '/app/autogpt/auto_gpt_workspace/cross_reference_weather_data.py': [Errno 2] No such file or directory
WEATHERGPT THOUGHTS:  I need to create a Python script to cross-reference the weather data from both sources. I will create a new file and write the necessary code to cross-reference the data.
REASONING:  Creating a Python script will allow me to automate the process of cross-referencing the weather data from both sources.
PLAN:
-  Create a new Python file.
-  Write the necessary code to cross-reference the weather data from both sources.
-  Save the file.
CRITICISM:  I need to ensure that the code I write is efficient and accurate.
SPEAK:  I need to create a Python script to cross-reference the weather data from both sources. Please wait a moment while I create the necessary file and write the code.
```

File: autogpt/commands/execute_code.py [hunk 1]
```diff
@@ -21,17 +22,25 @@ def execute_python_file(filename: str, config: Config) -> str:
...
+    )
 
     if not filename.endswith(".py"):
         return "Error: Invalid file type. Only .py files are allowed."
 
     if not os.path.isfile(filename):
-        return f"Error: File '{filename}' does not exist."
+        # Mimic the response that you get from the command line so that it's easier to identify
+        return (
+            f"python: can't open file '{filename}': [Errno 2] No such file or directory"
```
Comment (Pwuts) [inline autogpt/commands/execute_code.py:39]: This is a good idea

File: autogpt/commands/execute_code.py [hunk 2]
```diff
@@ -21,17 +22,25 @@ def execute_python_file(filename: str, config: Config) -> str:
...
     """
-    logger.info(f"Executing file '{filename}'")
+    logger.info(
+        f"Executing python file '{filename}' in working directory '{CFG.workspace_path}'"
+    )
 
     if not filename.endswith(".py"):
         return "Error: Invalid file type. Only .py files are allowed."
 
     if not os.path.isfile(filename):
```
Comment (Pwuts) [inline autogpt/commands/execute_code.py]: Nit: this does not enforce `restrict_to_workspace`. Would you mind amending it to use `Workspace.get_path`?

File: autogpt/commands/file_operations_utils.py [hunk 3]
```diff
@@ -146,7 +146,7 @@ def is_file_binary_fn(file_path: str):
 
 def read_textual_file(file_path: str, logger: logs.Logger) -> str:
     if not os.path.isfile(file_path):
-        raise FileNotFoundError(f"{file_path} not found!")
+        raise FileNotFoundError(f"read_file: {file_path}: no such file or directory")
```
Comment (Pwuts) [inline autogpt/commands/file_operations_utils.py]: Nit: that's a lot of colons, doesn't help with readability

(note that the error message is prepended with `Error:` as well)

**Review by Pwuts — COMMENTED** [review]
Thanks for picking this up!

File: autogpt/commands/file_operations_utils.py (hunk 3 above)
Comment (erik-megarad) [reply autogpt/commands/file_operations_utils.py]: I'm copying how the command line works:

```
$ cat sfasdf
cat: sfasdf: No such file or directory
```

File: autogpt/commands/execute_code.py (hunk 2 above)
Comment (erik-megarad) [reply autogpt/commands/execute_code.py]: Oh, this is code that already existed. I didn't change it on purpose to limit regressions.

File: autogpt/commands/file_operations_utils.py (hunk 3 above)
Comment (erik-megarad) [reply autogpt/commands/file_operations_utils.py]: Or, to be clear: I agree with you. I would prefer only one colon. I guess that decision was made like 30 years ago, oh well.

Comment (Pwuts) [reply autogpt/commands/file_operations_utils.py]: The example you give results in the message `Error: cat: sfasdf: No such file or directory`. Three colons is too much for me to parse at a glance.
```suggestion
        raise FileNotFoundError(f"read_file {file_path} failed: no such file or directory")
```

File: autogpt/commands/execute_code.py (hunk 2 above)
Comment (Pwuts) [reply autogpt/commands/execute_code.py]: I know it already existed. It should be fixed regardless, and I think it fits within the scope given by the title of this PR. :)

Comment (erik-megarad) [reply autogpt/commands/execute_code.py]: OK I added it. Not sure if the way I instantiated a `Workspace` is correct. Would be nice to be able to grab the agent's workspace.

**Review by Pwuts — APPROVED** [review]

**Comment by maxfield-allison** [comment]
Thank you for your service!

**Pull request merged**